scvi/inference/posterior.py

```diff
@@ -643,47 +705,149 @@ def entropy(hist_data):
+    idx_other = (all_labels == other_cell_idx) if other_cell_idx is not None else (
+            all_labels != cell_idx)
     if genes_idx is not None:
         px_scale = px_scale[:, genes_idx]
-    sample_rate_a = px_scale[idx].reshape(-1, px_scale.shape[1])
-    sample_rate_b = px_scale[idx_other].reshape(-1, px_scale.shape[1])
 
-    # agregate dataset
-    samples = np.vstack((sample_rate_a, sample_rate_b))
-
-    if sample_pairs is True:
-        # prepare the pairs for sampling
-        list_1 = list(np.arange(sample_rate_a.shape[0]))
-        list_2 = list(sample_rate_a.shape[0] + np.arange(sample_rate_b.shape[0]))
-        if not permutation:
-            # case1: no permutation, sample from A and then from B
-            u, v = np.random.choice(list_1, size=M_permutation), np.random.choice(list_2, size=M_permutation)
-        else:
-            # case2: permutation, sample from A+B twice
-            u, v = (np.random.choice(list_1 + list_2, size=M_permutation),
-                    np.random.choice(list_1 + list_2, size=M_permutation))
+    # first extract the data
+    # Assert that at this point we no longer have batch dimensions
+    assert len(px_scale.shape) == 2
+    sample_rate_a = px_scale[idx, :]
+    sample_rate_b = px_scale[idx_other, :]
+    # sample_rate_a = px_scale[:, idx, :]
+    # sample_rate_b = px_scale[:, idx_other, :]
+    sample_rate_a = sample_rate_a.reshape((-1, px_scale.shape[-1]))
+    sample_rate_b = sample_rate_b.reshape((-1, px_scale.shape[-1]))
+    samples = np.concatenate((sample_rate_a, sample_rate_b), axis=0)
+    # prepare the pairs for sampling
+    list_1 = list(np.arange(sample_rate_a.shape[0]))
+    list_2 = list(sample_rate_a.shape[0] + np.arange(sample_rate_b.shape[0]))
+
+    if importance_sampling:
+        weight_a = log_ratios[:, idx]
+        weight_b = log_ratios[:, idx_other]
+
+        # second let's normalize the weights
+        weight_a = softmax(weight_a, axis=0)
+        weight_b = softmax(weight_b, axis=0)
+        # reshape and aggregate dataset
+        weight_a = weight_a.flatten()
+        weight_b = weight_b.flatten()
+        weights = np.concatenate((weight_a, weight_b))
+
+        probas_a = weight_a / np.sum(idx)
+        probas_b = weight_b / np.sum(idx_other)
+
+        u, v = get_sampling_pair_idx(list_1, list_2, do_sample=sample_pairs,
+                                     permutation=permutation, M_permutation=M_permutation,
+                                     probas_a=probas_a, probas_b=probas_b)
 
         # then constitutes the pairs
-        first_set = samples[u]
-        second_set = samples[v]
+        first_samples = samples[u]
+        second_samples = samples[v]
+        first_weights = weights[u]
+        second_weights = weights[v]
+        to_sum = first_weights[:, np.newaxis] * second_weights[:, np.newaxis] * (
+                first_samples >= second_samples)
+        incomplete_weights = first_weights * second_weights
```

Comment:
I had some problem in practise with the imcomplete weights which would sometimes sum to zero. So it seems like using importance sampling means subselecting pairs of cells and using all the importance samples to compute the Bayes factor. I tried here to subsample from both the cells and the importances samples weight but it is really annoying in practise.